### PR TITLE
[BI-1056] GET germplasm override

### PR DESCRIPTION
**Pull request by ctucker3**
# Description
**Story:** https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?modal=detail&selectedIssue=BI-1056

Overrides the brapi proxy for GET germplasm. Following changes to the code were made:
- Assigns breedingMethodDbId the value of additionalInfo->breedingMethodId
- Removes [\<program key\>-\<accessionNumber\>] from the germplasm name
- Changes pedigree string from \<female parent name\>/\<male parent name\> to \<female parent GID\>/\<male parent GID\>
- Fixes a small pagination bug 
  - If pageSize was 20, but there were only 4 results, pageSize changed to 4, but should remain 20.
- Fixes sorting on the backend so it is now alphanumeric.
- Added tests for GET germplasm
- Added sorting to brapi query. 
- Added default GID sorting to GET germplasm.

# Dependencies
sgn: BI-1056 (https://github.com/Breeding-Insight/sgn/pull/50)
bi-api: BI-1056 (https://github.com/Breeding-Insight/bi-api/pull/130)
bi-web: BI-1056 (https://github.com/Breeding-Insight/bi-web/pull/165)

# Testing
See bi-web: BI-1056 for testing details.

# TAF Results
See bi-web: BI-1056 for details.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [x] I have create/modified unit tests to cover this change
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [x] I have run TAF.

### CODE REVIEW

**Review by timparsons — CHANGES_REQUESTED** [review]

File: src/main/java/org/breedinginsight/brapi/v2/GermplasmController.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    @ProgramSecured(roleGroups = {ProgramSecuredRoleGroup.ALL})
+    public HttpResponse<String> getGermplasm(
+            @PathVariable("programId") UUID programId,
+            @QueryValue @QueryValid(using = GermplasmQueryMapper.class) @Valid BrapiQuery queryParams) {
+        List<BrAPIGermplasm> germplasm = germplasmService.getGermplasm(programId);
+        queryParams.setSortField(germplasmQueryMapper.getDefaultSortField());
+        queryParams.setSortOrder(germplasmQueryMapper.getDefaultSortOrder());
+        Response<DataResponse<Object>> response = ResponseUtils.getBrapiQueryResponse(germplasm, germplasmQueryMapper, queryParams).body();
+        // Jackson can't serialize the additional info
+        String respBody = gson.toJson(response);
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/brapi/v2/GermplasmController.java]: Added a JsonSerializer

File: src/main/java/org/breedinginsight/brapi/v2/model/response/mappers/GermplasmQueryMapper.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public boolean exists(String fieldName) {
+        return getFields().containsKey(fieldName);
+    }
+
+    @Override
+    public Function<BrAPIGermplasm, ?> getField(String fieldName) throws NullPointerException {
+        if (fields.containsKey(fieldName)) return fields.get(fieldName);
+        else throw new NullPointerException();
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/brapi/v2/model/response/mappers/GermplasmQueryMapper.java]: Changed

File: src/main/java/org/breedinginsight/brapi/v2/services/BrAPIGermplasmService.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        germplasmSearch.externalReferenceSources(Arrays.asList(String.format("%s/programs", referenceSource)));
+        List<BrAPIGermplasm> germplasmList;
+        try {
+            germplasmList = BrAPIDAOUtil.search(
+                    api::searchGermplasmPost,
+                    api::searchGermplasmSearchResultsDbIdGet,
+                    germplasmSearch
+            );
+        } catch (ApiException e) {
+            throw new InternalServerException(e.getMessage());
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/brapi/v2/services/BrAPIGermplasmService.java]: Changed

File: src/main/java/org/breedinginsight/brapi/v2/GermplasmController.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+import java.util.UUID;
+
+@Slf4j
+@Controller
+@Secured(SecurityRule.IS_AUTHENTICATED)
+public class GermplasmController {
+
+    private BrAPIGermplasmService germplasmService;
+    private Gson gson;
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/brapi/v2/GermplasmController.java:33]: I don't think this is needed anymore

**Review by timparsons — CHANGES_REQUESTED** [review]

File: src/main/java/org/breedinginsight/brapi/v2/GermplasmController.java (hunk 4 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/brapi/v2/GermplasmController.java]: removed

**Review by timparsons — APPROVED** [review]

**Review by davedrp — APPROVED** [review]

**Pull request merged**